sem: rework explicit generic calls

zerbina opened the pull request:
## Summary

Rework and unify the handling of calls with explicitly provided type bindings (e.g.: `someRoutine[int, float]()`), significantly improving the usability of generic routines and fixing a large amount of related issues/bugs.

The core of the change is, that generic routines are now, except if explicitly requested otherwise, instantiated *after* overload resolution, instead of *before*. Forcing early instantiation is still possible via the `(someRoutine[int, float])()` syntax.

Since `void` parameter erasure only applies when when using early instantiation, code depending on the erasure needs to be adjusted.

Language changes/clarifications
-------------------------------

- default values for generic parameters on routines are now supported
- alias macros/templates are no longer considered for expansion in the callee position of the `someRoutine[A, B](c)` construct, making the behaviour consistent with that of normal call expressions
- the `someRoutine[]()` syntax is now valid. Instead of being treated as a dereference operation, it can now be used to force overload resolution to pick a generic routine (which includes nullary generics)
- using parenthesized callee expressions is now supported when the callee evaluates to a symbol choice
- the `T = x` (`nkExprEqExpr`) syntax is disallowed for type arguments

Fixed issues
------------

- fix passing values to type arguments with the `static[T]` not working when the callee is a template/macro, or the call located inside a template body / generic routine. A `type expected, but got` error was previously reported when attempting to pass the type argument
- all generic routines are now considered during overload resolution, regardless of which routine lookup finds first

## Details

### Previous implementation

Invocation of a routine where explicit generic type arguments where provided was handled by the compiler in two different ways:
1. explicit instantiation + overload resolution using the resulting instantiations (also referred to as early or eager instantiation here)
2. normal overload resolution (also referred to as late instantiation here)

How generic calls were treated was very dependent on the first symbol found by lookup. Early instantiation instantiation happened when:
- the first routine found by qualified lookup is not a template or macro
- the first non-template/macro routine in the candidate set had implicit generic parameters, and none of them were explicitly supplied
- the call expression was located in either the body of a template or generic routine (macros excluded)
while normal overload resolution was chosen in all other cases.

The early instantiation route worked by: gathering a set with all overloads; discarding all templates, macros, and routines with a different number of expected generic parameters (both explicit and implicit) than provided generic arguments; and generating an instantiation
for each remaining routine where the generic arguments matched the formal types.

If the resulting set was empty, a `cannot instantiate` error was produced. Otherwise, the closed set with the instantiated routines was passed on to normal overload resolution.

When taking the late instantiation route, the call expression with the callee being an `nkBracketExpr` node is passed directly to `resolveOverloads`, where the expressions in the bracket are analysed as normal operand expressions and the *types* of the analysed expressions are used as the initial bindings. No arity nor formal vs. argument type checking took place in this case, which allowed for semantically invalid programs to compile and possibly run.

There was another quirk with the late instantiation route: if the first found symbol is a macro or template, or the call expression is coming from a template/generic instantiation macros, each expression in the brackets was required to be a type expression (which prevents most valid `static` expressions), making it effectively impossible to use routines with explicit generic parameters using `static[T]` in these contexts.

### Now

* analyse and resolve all call expressions of the exact form `someRoutine[A, B](c)` through normal overload resolution
* extend `sigmatch` to perform the matching of generic arguments (if present) against the generic parameters -- a mismatch results in the overload to be rejected
* use the same analysis for generic arguments in the context of explicit instantiation (`explicitInstantiation`) and overload resolution (`resolveOverloads`)
* use dedicated processing for the generic call syntax in `semIndirectOp`, which allows for `someRoutine[]()` to work and also prevents alias templates/macros to be expanded in the callee symbols slot

### Other changes

* an implicit dereference is now inserted for `x.field()` where `field` is a view, making the behaviour of `nkDotExpr` consistent with other expressions in the context of indirect calls

### `void` parameter erasure

`sigmatch` could be adjusted to treat `void` parameters as effectively non-existing, though this wouldn't make `proc` types using generic parameters that are bound to `void` work.

### Error messages

When not explicitly requested, no instantiation of generic routines takes place prior to overload resolution, meaning that, for example, a type argument mismatch doesn't produce a `cannot instantiate` error anymore.

The `MismatchInfo` is extended to also include information about generic-parameter-related errors, but rendering of the error is still very basic.

---
<!-- Note: section break (`---`) onwards is not in CI merge commit -->

## To-do
- [X] ~match default parameter values against the formal type~
- [X] add regression tests for the fixed issues

## Notes for Reviewers
* improving the error messages makes more sense as separate PR, I think
* the changeset is quite small compared to its impact
* I'm not sure if the language changes / additions should be reflected in the manual yet

<!--
Pull Request(PR) Help

Before Merge Ensure:
* title reads like a short changelog line entry
* code includes tests and is documented
* leave the source better than before, but split out big reformats

See contributor (guide)[https://nim-works.github.io/nimskull/contributing.html]
for details, especially if you're new to this project.

Tips that make PRs easier:
* for big/impactful changes, start with chat/discussions to refine ideas
* refine the pull request message over time; don't have to nail it in one go
* handle the single commit message requirement at the end of review

saem commented on compiler/sem/semcall.nim:
```diff
@@ -283,34 +279,82 @@ proc getMsgDiagnostic(
+
+  var hasError = false
+  for i in 1..<n.len:
+    # a generic argument must be either evaluate to a type (``tyTypedesc``) or
```
```suggestion
    # a generic argument must either evaluate to a type (``tyTypedesc``) or
```

saem commented on compiler/sem/semcall.nim:
```diff
@@ -283,34 +279,82 @@ proc getMsgDiagnostic(
+      f = args
     else:
-      f = f[0]
+      n[0] = args
```
I wonder if manipulating `n[0]` like this may result in overloaded call operator macros/templates not being able to receive the callee name position as untyped?

Edit: Oh wait... I'm basing this off my older understanding, perhaps the paths into here are guarded for that.

zerbina replied:
All paths that reach this mutation don't lead back to the call operator application, so these are not affected.

A callee bracket expression that represents the passed type arguments in the context of a generic routine call now redirects to `semDirectOp` in `semIndirectOp`.

Nonetheless, the mutation of input AST is a problem (although pre-existing; see `semOpAux`). Two possible solutions:
1. `copyTree` the call expression at the start of `resolveOverloads`
2. store the analysed operands in a separate list (or `PNode`) that is passed to `sigmatch`

saem replied:
No action for this comment, thanks for clarifying.

---

I wish I had an easier time teaching that/keeping the fan in/out of these recursive calls straight in my head. There is some sort of memory technique I'm missing about how to easily build up/hold that picture in my head.

---

Tangent: there is something pretty interesting about option 2 (separate analysed operand storage), but I think it needs an analysis level marker on the node. It would tells us that certain parts of `sons` aren't part of the AST if we raise the node level back up. At which point we could have "metadata" in certain node positions. 

Alternatively, we have a new node kind (maybe `nkArgument`) that can hold this stuff in various positions. It could be nicely sanitized by extracting the nodes from a dedicated untyped sons position.

saem commented on compiler/sem/semcall.nim:
```diff
@@ -429,19 +473,33 @@ proc inferWithMetatype(c: PContext, formal: PType,
+        of nkEmpty:
+          # ``instantiateProcType`` uses ``nkEmpty`` to signal an incompatible
+          # default expression
+          conf.newError(calleeParams[i],
```
```suggestion
          hasError = true
          conf.newError(calleeParams[i],
```

I might have missed it, but I think that's required.

zerbina replied:
Yep, it's required.

saem reviewed (commented):
Overall, the intended semantics as laid out in the PR descriptin jives with the picture in my head of how things should work. The void parameter handling and template/macro expansion in callee symbol positions are also the right call for the purposes of this PR.

I did an initial pass of the code. There is a potential with some AST manipulation, but I might not be following the code correctly. This should surface in the call operator tests (`tests/specialops/tcallops` and friends).

I need to run through the code again to provide more insightful feedback.

saem commented:
Marking `tparamsindefault` as a known issue is a reasonable approach.

Alternatively, a fix is also possible, in `semtypes`, in the `semProcTypeNode -> semParType -> semTypeNode -> nkSym branch` chain. In there, when the symbol is one of `skParam/Local/Let/Var/etc`, basically any storage, then we're taking a compile time type value and moving it to type land. Here we can unwrap one level of `typedesc` to determine the type. This logic also applies to the `nkIdent/nkAccQuoted` branch.

saem commented on compiler/sem/semexprs.nim:
```diff
@@ -2048,6 +2048,9 @@ proc semSubscript(c: PContext, n: PNode, flags: TExprFlags): PNode =
             elif n[0].kind in nkSymChoices: n[0][0].sym
             else: nil
     if s != nil:
+      # XXX: how overloadable symbols are handled here is problematic, as what
```
thanks for calling this out, it's a good reminder.

saem commented on tests/lang_callable/generics/tlate_bound_typedesc.nim:
```diff
@@ -0,0 +1,15 @@
+
+# original issue https://github.com/nim-lang/nim/issues/11660
+
+# XXX: this semantics of ``typedesc`` as the parameter type are very likely
```
This is a good description of the issue, and when the behaviour is fixed the test will be a good reminder.

saem reviewed (approved):
Outside of any quibbles from CI, this is good to go: 🚀

zerbina commented:
The `tv2_cast.nim` test previously failed because the order of the definitions changed. This is because `injectdestructors` iterates over an unordered table that uses symbol-IDs as part of the key, when moving `def` statements. With less instantiations (and thus symbols) being generated now, the order of the table items changed. 

I adjusted the test for now, but a follow-up PR making the pass' output symbol-ID-independent is required. Note that only the order of the definition statements is affected, but not the order in which destructors are injected, meaning that the issue is (currently) purely cosmetic.

zerbina committed:
sem: rework explicit generic calls

Summary
=======

Rework and unify the handling of calls with explicitly provided type
bindings (e.g.: `someRoutine[int, float]()`), significantly improving
the usability of generic routines and fixing a large amount of related
issues/bugs.

The core of the change is, that generic routines are now instantiated
*after* overload resolution, instead of *before*. Forcing early
instantiation is still possible via the `(someRoutine[int, float])()`
syntax.

Since `void` parameter erasure only applies when using early
instantiation, code depending on the erasure needs to be adjusted.

Language changes/clarifications
-------------------------------

- default values for generic parameters are now supported
- alias macros/templates are no longer considered for expansion in the
  callee position of the `someRoutine[A, B](c)` construct, making the
  behaviour consistent with that of normal call expressions
- the `someRoutine[]()` syntax is now valid. Instead of being treated
  as a dereference operation, it can now be used to force overload
  resolution to pick a generic routine (which includes nullary
  generics)
- parenthesized callee expressions that evaluate to symbol choices are
  legal and now supported
- the `T = x` (`nkExprEqExpr`) syntax is disallowed for type arguments

Fixed issues
------------

- fix passing values to type arguments with the `static[T]` not working
  when the callee is a template/macro, or the call located inside a
  template body / generic routine. A `type expected, but got` error was
  previously reported when attempting to pass the type argument
- all generic routines are now considered during overload resolution,
  regardless of which routine lookup finds first

Details
=======

Previous implementation
-----------------------

Invocation of a routine where explicit generic type arguments where
provided was handled by the compiler in two different ways:

1. explicit instantiation + overload resolution using the resulting
   instantiations (also referred to as early or eager instantiation
   here)
2. normal overload resolution (also referred to as late instantiation
   here)

How generic calls were treated was very dependent on the first symbol
found by lookup. Early instantiation instantiation happened when:
- the first routine found by qualified lookup is not a template or macro
- the first non-template/macro routine in the candidate set had implicit
  generic parameters, and none of them were explicitly supplied
- the call expression was located in either the body of a template or
  generic routine (macros excluded)
while normal overload resolution was chosen in all other cases.

The early instantiation route worked by: gathering a set with all
overloads; discarding all templates, macros, and routines with a
different number of expected generic parameters (both explicit and
implicit) than provided generic arguments; and generating an
instantiation for each remaining routine where the generic arguments
matched the formal types.

If the resulting set was empty, a `cannot instantiate` error was
produced. Otherwise, the closed set with the instantiated routines was
passed on to normal overload resolution.

When taking the late instantiation route, the call expression with the
callee being an `nkBracketExpr` node is passed directly to
`resolveOverloads`, where the expressions in the bracket are analysed
as normal operand expressions and the *types* of the analysed
expressions are used as the initial bindings. No arity nor formal vs.
argument type checking took place in this case, which allowed for
semantically invalid programs to compile and possibly run.

There was another quirk with the late instantiation route: if the first
found symbol is a macro or template, or the call overloaded and coming
from a template/generic instantiation macros, each expression in the
brackets was required to be a type expression parseable by `semTypeNode`
(which rejects most expressions not evaluating to `typedesc`), making
it effectively impossible to use routines with explicit `static[T]` type
parameters in these contexts.

Now
---

* analyse and resolve all call expressions of the exact form
  `someRoutine[A, B](c)` through normal overload resolution
* extend `sigmatch` to perform the matching of generic arguments
  (if present) against the generic parameters -- a mismatch results in
  the overload to be rejected
* use the same analysis for generic arguments in the context of
  explicit instantiation (`explicitInstantiation`) and overload
  resolution (`resolveOverloads`)
* use dedicated processing for the generic call syntax in
  `semIndirectOp`, which allows for `someRoutine[]()` to work and also
  prevents alias templates/macros to be expanded in the callee symbols'
  slot

Other changes
-------------

* an implicit dereference is now inserted for `x.field()` where `field`
  is a view, making the behaviour of `nkDotExpr` consistent with other
  expressions in the context of indirect calls

`void` parameter erasure
------------------------

`sigmatch` could be adjusted to treat `void` parameters as effectively
non-existing, though this wouldn't make `proc` types using generic
parameters that are bound to `void` work.

Error messages
--------------

When not explicitly requested, no instantiation of generic routines
takes place prior to overload resolution, meaning that, for example, a
type argument mismatch doesn't produce a `cannot instantiate` error
anymore.

The `MismatchInfo` is extended to also include information about
generic-parameter-related errors, but rendering of the error is still
very basic.

zerbina commented:
The PR is squashed and rebased -- I also added two more regression tests. If CI doesn't report any errors, I'm going to go ahead and merge it.

saem commented:
Looks to be spurious, re-running the job.

saem commented:
bors r+
